lib/workers/repository/updates/generate.ts

```diff
@@ -167,54 +168,23 @@ export function generateBranchConfig(
     } else if (semver.valid(toVersions[0])) {
       upgrade.isRange = false;
     }
-    // Use templates to generate strings
-    if (upgrade.semanticCommits === 'enabled' && !upgrade.commitMessagePrefix) {
+    const commitMessageFactory = new UpdateCommitMessageFactory(upgrade);
+    const commitMessage = commitMessageFactory.create();
+    if (
+      SemanticCommitMessage.is(commitMessage) &&
+      upgrade.semanticCommitScope
+    ) {
       logger.trace('Upgrade has semantic commits enabled');
-      let semanticPrefix = upgrade.semanticCommitType;
-      if (upgrade.semanticCommitScope) {
-        semanticPrefix += `(${template.compile(
-          upgrade.semanticCommitScope,
-          upgrade
-        )})`;
-      }
-      upgrade.commitMessagePrefix = CommitMessage.formatPrefix(semanticPrefix!);
-      upgrade.toLowerCase =
-        regEx(/[A-Z]/).exec(upgrade.semanticCommitType!) === null &&
-        !upgrade.semanticCommitType!.startsWith(':');
     }
-    // Compile a few times in case there are nested templates
-    upgrade.commitMessage = template.compile(
-      upgrade.commitMessage ?? '',
-      upgrade
-    );
-    upgrade.commitMessage = template.compile(upgrade.commitMessage, upgrade);
-    upgrade.commitMessage = template.compile(upgrade.commitMessage, upgrade);
     // istanbul ignore if
-    if (upgrade.commitMessage !== sanitize(upgrade.commitMessage)) {
+    if (commitMessage.subject !== sanitize(commitMessage.subject)) {
       logger.debug(
         { branchName: config.branchName },
         'Secrets exposed in commit message'
       );
       throw new Error(CONFIG_SECRETS_EXPOSED);
     }
-    upgrade.commitMessage = upgrade.commitMessage.trim(); // Trim exterior whitespace
-    upgrade.commitMessage = upgrade.commitMessage.replace(regEx(/\s+/g), ' '); // Trim extra whitespace inside string
-    upgrade.commitMessage = upgrade.commitMessage.replace(
-      regEx(/to vv(\d)/),
-      'to v$1'
-    );
-    if (upgrade.toLowerCase) {
-      // We only need to lowercase the first line
-      const splitMessage = upgrade.commitMessage.split(newlineRegex);
-      splitMessage[0] = splitMessage[0].toLowerCase();
-      upgrade.commitMessage = splitMessage.join('\n');
-    }
```

Comment:
> I still need to understand quite a lot about the code, but I think [this part](https://github.com/renovatebot/renovate/blob/bba3c782e257682feeb42ccb4d35f6db37c9e611/lib/workers/repository/updates/generate.ts#L181-L183) is the issue, but simply removing it could be a breaking change for others since all semantic commit messages would suddenly start with an uppercase letter in the subject.

Do you maintainers think this is a breaking change, and that we should put this in a new `major` release of Renovate bot?

Reply:
The uppercasing of the subject is not a problem in this solution anymore, but this will cause migration and onboarding commit messages to drop the ":" separator when not using semantic commit messages (which was an inconsistency between updates and migration and onboarding).

Reply:
I think we can push this to v33 to be safe, but we should try to understand exactly what's changed